### PR TITLE
Fix Asynchronous Counters Recording

### LINKED ISSUE #3278: AsyncCounter and AsyncUpDownCounter expected callback value changed in v0.32.0
Before 0.32, the value in a callback function was expected to be an absolute value, but since 0.32 it must be a delta or metrics are incorrect:

```go
hitsCounter, _ := meter.AsyncInt64().Counter("some.prefix.cache_hits")

// before 0.32
hitsCounter.Observe(ctx, hitsValue) // abs value

// 0.32
hitsCounter.Observe(ctx, hitsValue-prevHitsValue) // delta value
```

Is that intentional? It is not documented anywhere and not mentioned in the changelog. Also, [runtimemetrics](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/instrumentation/runtime/v0.36.3/instrumentation/runtime/runtime.go#L148-L162) still uses the old way to report observable values.

### PR DESCRIPTION
**Pull request by MrAlias**
Fix #3278

Currently, asynchronous counters observe values that are treated as incremental changes to already summed recordings. This is contrary to [what the specification notes](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.14.0/specification/metrics/api.md#asynchronous-counter-creation) is expected by the user in the API:

> Note: Unlike Counter.Add() which takes the increment/delta value, the callback function reports the absolute value of the counter. To determine the reported rate the counter is changing, the difference between successive measurements is used.

This fixes the implemented behavior. It adds a new `precomputedSum` `Aggregator` type that interprets the passed `Aggregate` values as already computed sums and returns them directly. This new `Aggregator` type is used for the `AsyncCounter` and `AsyncUpDownCounter` instrument kinds.

### CODE REVIEW

**Review by dashpole — APPROVED** [review]

File: sdk/metric/internal/sum.go [hunk 1]
```diff
@@ -151,3 +165,41 @@ func (s *cumulativeSum[N]) Aggregation() metricdata.Aggregation {
...
+// attributes and the aggregation cycle the measurements were made in.
+//
+// The monotonic value is used to communicate the produced Aggregation is
+// monotonic or not. The returned Aggregator does not make any guarantees this
+// value is accurate. It is up to the caller to ensure it.
+//
+// The temporality value is used to communicate the produced Aggregation
+// temporality. The returned Aggregator does not make any guarantees this value
+// is accurate. It is up to the caller to ensure it.
+func NewPrecomputedSum[N int64 | float64](monotonic bool, temporality metricdata.Temporality) Aggregator[N] {
```
Comment (MadVikingGod) [inline sdk/metric/internal/sum.go:180]: Is there a reason that temporality is a parameter here, unlike the other Sum Aggregators?

**Review by MadVikingGod — APPROVED** [review]
This seems reasonable.

Just for my understanding does this basically mean that async Counters are gauges with monotonic flags and temporality fields?

File: sdk/metric/internal/sum.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/internal/sum.go:180]: no, I'll update.

Comment (MrAlias) [reply sdk/metric/internal/sum.go]: updated.

**Comment by MrAlias** [comment]
> Just for my understanding does this basically mean that async Counters are gauges with monotonic flags and temporality fields?

I think the `AsyncUpDownCounter` could be though of this abstractly. It is still exported as a `Sum` data type, by default, in the end though.

The way I understand the specification to be written is that the user of asynchronous up-down-counters and counters maintain the state of the counter themselves. It is passed to the SDK with the appropriate monotonic and temporality types to have the SDK report the sum with the right properties, but it is up to the user to ensure the validity of these properties.

**Pull request merged**